src/FsAutoComplete.Core/Commands.fs

```diff
@@ -26,54 +27,67 @@ type Commands (serialize : Serializer) =
     member private x.SerializeResult (successToString: Serializer -> 'a -> string, ?failureToString: Serializer -> string -> string) =
         x.SerializeResultAsync ((fun s x -> successToString s x |> async.Return), ?failureToString = failureToString)
 
-    member __.TryGetRecentTypeCheckResultsForFile = checker.TryGetRecentCheckResultsForFile
-    member __.TryGetFileCheckerOptionsWithLinesAndLineStr = state.TryGetFileCheckerOptionsWithLinesAndLineStr
-    member __.TryGetFileCheckerOptionsWithLines = state.TryGetFileCheckerOptionsWithLines
-    member __.Files = state.Files
+    member private x.AsCancellable (filename : SourceFilePath) (action : Async<string list>) =
+        let cts = new CancellationTokenSource()
+        state.AddCancellationToken(filename, cts)
+        Async.StartCatchCancellation(action, cts.Token)
+        |> Async.Catch
+        |> Async.map (function Choice1Of2 res -> res | Choice2Of2 _ -> [Response.info serialize "Request cancelled"])
 
-    member __.Parse file lines version = async {
-        let colorizations = state.ColorizationOutput
-        let parse' fileName text options =
-            async {
-                let! result = checker.ParseAndCheckFileInProject(fileName, version, text, options)
-                return
-                    match result with
-                    | Failure e -> [Response.error serialize e]
-                    | Success (parseResult, checkResults) ->
-                        match checkResults with
-                        | FSharpCheckFileAnswer.Aborted -> [Response.info serialize "Parse aborted"]
-                        | FSharpCheckFileAnswer.Succeeded results ->
-                            let errors = Array.append results.Errors parseResult.Errors
-                            if colorizations then
-                                [ Response.errors serialize (errors, fileName)
-                                  Response.colorizations serialize (results.GetSemanticClassification None) ]
-                            else [ Response.errors serialize (errors, fileName) ]
-            }
+    member private x.CancelQueue (filename : SourceFilePath) =
+        state.GetCancellationTokens filename |> List.iter (fun cts -> cts.Cancel() )
+        state.ClearCancellationTokens filename
```

Comment:
Could be a race when a new Cancellation token arrives before `ClearCancellationTokens` and after `Cancelling` the existing ones.